src/wrapper/macros.rs

```diff
@@ -1,142 +1,91 @@
+        jni_call_check_ex!($jnienv, $version, $name $(, $args)*).and_then(|ret| {
+            if ret.is_null() {
+                Err($crate::errors::Error::NullPtr(concat!(stringify!($name), " result")))
+            } else {
+                Ok(ret)
             }
-            None => {
-                log::trace!("jnienv method not defined, returning error");
-                return Err($crate::errors::Error::JNIEnvMethodNotFound(stringify!(
-                    $name
-                )));
-            }
-        }
-    }};
-}
-
-macro_rules! check_exception {
-    ( $jnienv:expr ) => {
-        log::trace!("checking for exception");
-        let check = { jni_unchecked!($jnienv, ExceptionCheck) } == $crate::sys::JNI_TRUE;
-        if check {
-            log::trace!("exception found, returning error");
-            return Err($crate::errors::Error::JavaException);
-        }
-        log::trace!("no exception found");
-    };
-}
-
-macro_rules! catch {
-    ( move $b:block ) => {
-        (move || $b)()
-    };
-    ( $b:block ) => {
-        (|| $b)()
-    };
-}
-
-macro_rules! java_vm_unchecked {
-    ( $java_vm:expr, $name:tt $(, $args:expr )* ) => ({
-        log::trace!("calling unchecked JavaVM method: {}", stringify!($name));
-        java_vm_method!($java_vm, $name)($java_vm, $($args),*)
+        })
     })
 }
 
-macro_rules! java_vm_method {
-    ( $jnienv:expr, $name:tt ) => {{
-        log::trace!("looking up JavaVM method {}", stringify!($name));
-        let env = $jnienv;
-        match deref!(deref!(env, "JavaVM"), "*JavaVM").$name {
-            Some(meth) => {
-                log::trace!("found JavaVM method");
-                meth
-            }
-            None => {
-                log::trace!("JavaVM method not defined, returning error");
-                return Err($crate::errors::Error::JavaVMMethodNotFound(stringify!(
-                    $name
-                )));
-            }
+/// Calls a JNIEnv function, with no check for exceptions, then checks for a `null` return value
+///
+/// Returns `Err(Error::NullPtr)` if the JNI function returns `null`
+macro_rules! jni_call_only_check_null_ret {
+    ( $jnienv:expr, $version:tt, $name:tt $(, $args:expr )* ) => ({
+        let ret = jni_call_unchecked!($jnienv, $version, $name $(, $args)*);
+        if ret.is_null() {
+            Err($crate::errors::Error::NullPtr(concat!(stringify!($name), " result")))
+        } else {
+            Ok(ret)
         }
-    }};
+    })
 }
 
-macro_rules! deref {
+/// Maps a pointer to either Ok(ptr) or Err(Error::NullPtr)
+///
+/// This makes it reasonably ergonomic to use `?` to early-exit with an `Err` in
+/// case of `null` pointer arguments.
+///
+/// Unlike earlier macros this avoids using `return`, since that can result in
+/// surprising control flow if the caller doesn't realize that a macro might
+/// explicitly return from the current function.
+macro_rules! null_check {
```

Comment:
Currently it can't be a function because there's no trait that corresponds to the `is_null()` check for all the types that may be passed to `null_check`. We could potentially create one function that accepts an `AsRef<JObject<'a>>` but would then need separate APIs for non-object / raw pointer types. E.g it's also used like this currently:

```rust
pub unsafe fn from_raw(ptr: *mut sys::JavaVM) -> Result<Self> {
        let ptr = null_check(ptr, "from_raw ptr argument")?;
```